overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, fmt.Errorf("cannot create user: cannot get serial: %v", err)
+	}
+
+	db := assertstate.DB(st)
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  model,
+		serialAs: serial,
+		isSudoer: sudoer,
+	}
+
+	if email == "" {
+		return u.createAllKnownSystemUsers(st)
+	}
+
+	username, opts, err := getUserDetailsFromAssertion(u.assertDb, u.modelAs, u.serialAs, email)
+	if err != nil {
+		return nil, &UserError{Err: err}
+	}
+
+	opts.Sudoer = sudoer
+	createdUser, err := u.createUser(st, opts, username, email)
+	if err != nil {
+		return nil, err
+	}
+	return []CreatedUser{createdUser}, nil
+}
+
+// RemoveUser removes linux user account of passed username.
+func RemoveUser(st *state.State, username string) (*auth.UserState, error) {
+	// TODO: allow to remove user entries by email as well
+
+	// catch silly errors
+	if username == "" {
+		return nil, &UserError{Err: fmt.Errorf("need a username to remove")}
+	}
+
+	// check the user is known to snapd
+	_, err := auth.UserByUsername(st, username)
+	if err == auth.ErrInvalidUser {
+		return nil, &UserError{Err: fmt.Errorf("user %q is not known", username)}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// first remove the system user
+	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return nil, err
+	}
+
+	// then the UserState
+	u, err := auth.RemoveUserByUsername(st, username)
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, err
+	}
+	return u, nil
+}
+
+func getUserDetailsFromStore(theStore snapstate.StoreService, email string) (string, *osutil.AddUserOptions, error) {
+	v, err := theStore.UserInfo(email)
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot create user %q: %s", email, err)
+	}
+	if len(v.SSHKeys) == 0 {
+		return "", nil, fmt.Errorf("cannot create user for %q: no ssh keys found", email)
+	}
+
+	// Amend information where the key came from to ensure it can
+	// be update/replaced later
+	for i, k := range v.SSHKeys {
+		v.SSHKeys[i] = fmt.Sprintf(`%s # snapd {"origin":"store","email":%q}`, k, email)
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, v.OpenIDIdentifier)
+	opts := &osutil.AddUserOptions{
+		SSHKeys: v.SSHKeys,
+		Gecos:   gecos,
+	}
+	return v.Username, opts, nil
+}
+
+// createUserOpts is a helper to handle user operations
+type createUserOpts struct {
+	assertDb asserts.RODatabase
+	modelAs  *asserts.Model
+	serialAs *asserts.Serial
+	isSudoer bool
+}
+
+func (u *createUserOpts) createUser(st *state.State, opts *osutil.AddUserOptions, username, email string) (CreatedUser, error) {
+	opts.ExtraUsers = !release.OnClassic
+	createdUser, err := u.addUser(st, username, email, opts)
+	if err != nil {
+		return CreatedUser{}, err
+	}
+
+	return createdUser, nil
+}
+
+func (u *createUserOpts) createAllKnownSystemUsers(state *state.State) ([]CreatedUser, error) {
```

Comment:
It seems like this is the only method that actually uses the receiver's data. In that case, I would just pass the option's fields as parameters and delete the options alltogether. Maybe the DB object could even be created here instead of passed in from the exposed function.

Reply:
Nice! I've done all your comments in relation to this, I even removed the option structure.